internal/access/grant.go

```diff
@@ -29,30 +29,43 @@ func ListGrants(c *gin.Context, subject uid.PolymorphicID, resource string, priv
 
 	roles := []string{models.InfraAdminRole, models.InfraViewRole, models.InfraConnectorRole}
 	db, err := RequireInfraRole(c, roles...)
-	if err == nil {
-		selectors = append(selectors, data.ByOptionalSubject(subject))
-		return data.ListGrants(db, selectors...)
-	}
 	err = HandleAuthErr(err, "grants", "list", roles...)
-
 	if errors.Is(err, ErrNotAuthorized) {
 		// Allow an authenticated identity to view their own grants
-		db := getDB(c)
+		db = getDB(c)
 		subjectID, _ := subject.ID()
 		identity := AuthenticatedIdentity(c)
 		switch {
 		case identity == nil:
 			return nil, err
 		case subject.IsIdentity() && identity.ID == subjectID:
-			selectors = append(selectors, data.BySubject(subject))
+			if inherited {
+				selectors = append(selectors, data.GrantsInheritedBySubject(subject))
+			} else {
+				selectors = append(selectors, data.BySubject(subject))
+			}
```

Comment:
Are all 3 of these `if/else` blocks the same? I guess the last one is different in that it uses `ByOptionalSubject`.

What happens if `inherited` is used without a subject? I guess we should add no selector in that case? I think the `inherited && subject.IsIdentity()` you had originally was correct, the problem was only that it was not being applied in both authz scenarios. So maybe:

```go
if subject != "" {
    if inherited && subject.IsIdentity() {
        selectors = append(selectors, data.GrantsInheritedBySubject(subject))
    } else {
        selectors = append(selectors, data.BySubject(subject))
    }
}
```

I think if we did that once above, then we wouldn't need the `selectors = append(selectors, data.BySubject(subject))` we used to have here. We could perform the query with the already constructed list of selectors.

Reply:
that's not right because in some cases (requesting own grants) the subject is required. Merging this endpoint with the /users/x/grants endpoints added some of this complexity.

Reply:
The subject is required for authorization, but that check already exists.